loopy/transform/instruction.py

```diff
@@ -126,19 +126,46 @@ def add_dep(insn):
 def remove_instructions(kernel, insn_ids):
     """Return a new kernel with instructions in *insn_ids* removed.
 
-    Dependencies across (one, for now) deleted isntructions are propagated.
-    Behavior is undefined for now for chains of dependencies within the
-    set of deleted instructions.
+    Dependencies across deleted isntructions are recursively propagated.
 
     This also updates *no_sync_with* for all instructions.
+
+    :arg insn_ids: An instance of :class:`set` or :class:`str` as
+        understood by :func:`loopy.match.parse_match` or
+        :class:`loopy.match.MatchExpressionBase`.
     """
 
     if not insn_ids:
         return kernel
 
+    from loopy.match import MatchExpressionBase
+
+    if isinstance(insn_ids, str):
+        from loopy.match import parse_match
+        insn_ids = parse_match(insn_ids)
+    if isinstance(insn_ids, MatchExpressionBase):
+        within = insn_ids
+
+        insn_ids = set([insn.id for insn in kernel.instructions if
+            within(kernel, insn)])
+
     assert isinstance(insn_ids, set)
     id_to_insn = kernel.id_to_insn
 
+    # for each insn_id to be removed get deps in terms of remaining
+    # insns
+    retargeted_deps = {}
```

Comment:
Describe keys/values of this data structure.

Reply:
Done, see  f02638d.